### PR TITLE
Finalize LRU algorithm #226

### PR DESCRIPTION
**Pull request by sergiuilie**

### CODE REVIEW

**Comment by ishche** [comment]
Have you considered using the order instead of the time?

**Comment by sergiuilie** [comment]
@ishche I am using the hit and after the time, the time should also give the correct order as two files are created at separate moments in time, I am not sure I understand what do you mean by order, simply the order of the elements how they are stored in the cache?

File: com.ca.lsp.cobol/lsp-core-domain/src/main/java/com/broadcom/lsp/domain/cobol/databus/impl/CopybookRepositoryLRU.java [hunk 1]
```diff
@@ -84,6 +88,8 @@ public void setSort(boolean isSort) {
   @Override
   @SneakyThrows
   public synchronized void persist(@NonNull CopybookStorable deepCopy) {
+    cpyRepo.removeIf(elem -> elem.isExpired());
```
Comment (zacanbrcom) [inline com.ca.lsp.cobol/lsp-core-domain/src/main/java/com/broadcom/lsp/domain/cobol/databus/impl/CopybookRepositoryLRU.java]: We can use the method reference directly ```CopybookStorable::isExpired```

**Comment by ishche** [comment]
> @ishche I am using the hit and after the time, the time should also give the correct order as two files are created at separate moments in time, I am not sure I understand what do you mean by order, simply the order of the elements how they are stored in the cache?

By order for LRU(Least recently used) I meant the order based on recent use. It seems not very useful to relay on a timer.

**Comment by sergiuilie** [comment]
this is how I use, based on hit first and after time, the hit is the last recently used

**Comment by ishche** [comment]
> this is how I use, based on hit first and after time, the hit is the last recently used

Yes, but what is the value to have time involved?

**Comment by sergiuilie** [comment]
> > this is how I use, based on hit first and after time, the hit is the last recently used
> 
> Yes, but what is the value to have time involved?

We wanna get rid of the file which is expired, for now 3h, doesn't matter how used it is, and to order by hits and after by youngest

**Comment by zacanbrcom** [comment]
Guys, as general remarks I saw that we don't have unit test for negative scenarios, is something that we should fix (in this or another ad-hoc issue)

File: com.ca.lsp.cobol/lsp-core-domain/src/test/java/com/broadcom/lsp/domain/cobol/databus/impl/CopybookRepositoryLRUTest.java [hunk 2]
```diff
@@ -62,6 +66,60 @@ public void testPersist() {
...
+   * time, it is worth to mention that we need to put the thread asleep for 1ms in order to have a
+   * time difference at the creation moment, if not it will not differentiate and the order will be
+   * not the expected one
+   *
+   * @throws InterruptedException
+   */
+  @Test
+  public void testCacheSort() throws InterruptedException {
+    CopybookStorable topElem = new CopybookStorable("NEW_STO3", "URI", "DUMMY CONTENT");
+    Thread.sleep(1);
```
Comment (zacanbrcom) [inline com.ca.lsp.cobol/lsp-core-domain/src/test/java/com/broadcom/lsp/domain/cobol/databus/impl/CopybookRepositoryLRUTest.java]: Here is better use to use a waiter

**Comment by ishche** [comment]
> > > this is how I use, based on hit first and after time, the hit is the last recently used
> > 
> > 
> > Yes, but what is the value to have time involved?
> 
> We wanna get rid of the file which is expired, for now 3h, doesn't matter how used it is, and to order by hits and after by youngest

But you have all the necessary information to invalidate the cache. The LRU usually used to deal with out of space in the cache.
Well, If know what are you doing, I'm ok with this solution.

**Pull request merged**